### ISSUE TITLE
tradestatistics package

### ISSUE THREAD

**Issue by pachamaltese** [ropensci/software-review#274]
Submitting Author: Pachá (@pachamaltese)  
Repository: https://github.com/tradestatistics/tradestatistics
Version submitted: 0.1
Editor: @maelle   
Reviewer 1: TBD  
Reviewer 2: TBD  
Archive: TBD  
Version accepted: TBD   

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: tradestatistics
Type: Package
Title: Open Trade Statistics API Wrapper and Utility Program
Version: 0.1
Authors@R: c(
  person("Mauricio", "Vargas S.", role = c("aut", "cre", "cph"), email = "mvargas@dcc.uchile.cl"),
  person("Joshua", "Kunst", role = c("ctb")),
  person("Amanda", "Dobbyn", role = c("ctb")),
  person("Jorge", "Cimentada", role = c("ctb")),
  person(family = "UN Comtrade", role = c("dtc"))
  )
URL: https://CRAN.R-project.org/package=tradestatistics
BugReports: https://github.com/tradestatistics/ts-r-package/issues
Description: Access Open Trade Statistics API from R to download international trade data.
License: GPL-3
LazyData: TRUE
Depends:
  R (>= 3.2)
Imports:
  rlang,
  magrittr,
  dplyr,
  stringr,
  crul,
  purrr,
  jsonlite
RoxygenNote: 6.1.1
Suggests: 
  knitr,
  rmarkdown,
  testthat,
  covr,
  styler,
  pkgdown,
  vcr
VignetteBuilder: knitr
```

## Scope 

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [x] data retrieval
	- [ ] data extraction
	- [ ] database access
	- [ ] data munging
	- [ ] data deposition
	- [x] reproducibility
	- [ ] geospatial data
	- [ ] text analysis
	

- Explain how the and why the package falls under these categories (briefly, 1-2 sentences):

Because the package connects to an API and does API calls to simplify things for the final user who wants imports/exports and some metrics such as % of increase/decrease.

-   Who is the target audience and what are scientific applications of this package?  

Non-expert users that use international trade data. This can also be targeted to intermediate/advanced users who can benefit from the speed and short syntax that this package provides.

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

No, this is made for a new data project

-   If you made a pre-submission enquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

No.

## Technical checks

Confirm each of the following by checking the box.  This package:

- [x] does not violate the Terms of Service of any service it interacts with. 
- [x] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme). 
- [x] includes [documentation with examples for all functions](https://ropensci.github.io/dev_guide/building.html#documentation).
- [x] contains a vignette with examples of its essential functions and uses.
- [x] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [x] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage using services such as Travis CI, Coveralls and/or CodeCov.

## Publication options

- [x] Do you intend for this package to go on CRAN?  
- [ ] Do you wish to automatically submit to the [Journal of Open Source Software](http://joss.theoj.org/)? If so:

<details>
 <summary>JOSS Options</summary>  

  - [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements).
    - [ ] The package contains a `paper.md` matching [JOSS's requirements](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) with a high-level description in the package root or in `inst/`.
    - [ ] The package is deposited in a long-term repository with the DOI: 
    - (*Do not submit your package separately to JOSS*)  
  
</details>

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:  

<details>
<summary>MEE Options</summary>  

- [ ] The package is novel and will be of interest to the broad readership of the journal. 
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [ROpenSci's Code of Conduct](https://ropensci.github.io/dev_guide/policies.html#code-of-conduct) during the review process and in maintaining my package should it be accepted.

**Comment by maelle** [comment]
:wave: @pachamaltese! Thanks for your submission! Can you confirm the package and underlying data source are mature enough for review?

**Comment by pachamaltese** [comment]
@maelle Hi!! Yes, I added all the feedback from the previous package and I extended it to obtain data from different tables provided I built my own tailored API

**Comment by pachamaltese** [comment]
@maelle Hi! sorry, I totally forgot to mention that the ROpenSci badge that I added should get another numeral provided 217 is for the older package

**Comment by maelle** [comment]
### Editor checks:

- [x] **Fit**: The package meets criteria for [fit](policies.md#fit) and [overlap](policies.md#fit)
- [x] **Automated tests:** Package has a testing suite and is tested via Travis-CI or another CI service.
- [x] **License:** The package has a CRAN or OSI accepted license
- [x] **Repository:** The repository link resolves correctly
- [ ] **Archive** (JOSS only, may be post-review): The repository DOI resolves correctly
- [ ] **Version** (JOSS only, may be post-review): Does the release version given match the GitHub release (v1.0.0)?

---

#### Editor comments

Thanks for your submission! I'm now looking for reviewers. Below are a few comments/questions.

In the README please write GPL3 licence to be more specific.

Is it necessary for the repo to be a fork? Or can you at least add an issue tracker to it?

Could you fix the example in `get_countrycode()` docs or at least add another working example?

Also note that when transferring it to the ropensci GitHub organization, it'll need to have the same name as the package.

Please update the review badge, the number in it should be the issue number, 274.

Why is there a CRAN link in DESCRIPTION given the package has not been released on CRAN? By the way please do not submit the package to CRAN before the end of the review process.

There's no example after the sentence "Here’s a basic example to obtain bilateral trade between Chile and Argentina in the year 2000 under different trade classifications:" in the README.

Here's the output from `goodpractice`, for you and the reviewers to have in mind.

```
 ✖ avoid long code lines, it is bad for readability.
    Also, many people prefer editor windows that are about 80
    characters wide. Try make your lines shorter than 80 characters

    R/get_countrycode.R:60:1
    R/get_countrycode.R:67:1
    R/get_data.R:44:1
    R/get_data.R:47:1
    R/get_data.R:97:1
    ... and 15 more lines

  ✖ fix this R CMD check NOTE: Note: found 4 marked UTF-8
    strings
  ✖ fix this R CMD check WARNING: LaTeX errors when
    creating PDF version. This typically indicates Rd problems.
  ✖ fix this R CMD check ERROR: Re-running with no
    redirection of stdout/stderr. Hmm ... looks like a package You
    may want to clean up by 'rm -rf
    /tmp/RtmpYe7MiN/Rd2pdf33297bad420f'
```
Typos

```
cointain      basic-usage.Rmd:45
mesured       basic-usage.Rmd:114
perfoms       get_data.Rd:39
```
---

Reviewers:
Due date:

**Comment by pachamaltese** [comment]
Hi @maelle 
I tested on another laptop and I still cannot reproduce the UTF-8 warnings
Here's my output
```
> pkgdown::build_site()
══ Building pkgdown site ════════════════════════════════════════════════════════════════════════════════════
Reading from: '/home/pacha/github/tradestatistics'
Writing to:   '/home/pacha/github/tradestatistics/docs'
── Initialising site ────────────────────────────────────────────────────────────────────────────────────────
── Building home ────────────────────────────────────────────────────────────────────────────────────────────
Reading 'CODE_OF_CONDUCT.md'
Reading 'README.Rmd'
Writing 'index.html'
── Building function reference ──────────────────────────────────────────────────────────────────────────────
Updating tradestatistics documentation
Writing NAMESPACE
Loading tradestatistics
Writing NAMESPACE
Loading tradestatistics
Reading 'man/countries.Rd'
Reading 'man/get_countrycode.Rd'
Reading 'man/get_data.Rd'
Reading 'man/get_productcode.Rd'
Reading 'man/products.Rd'
Reading 'man/tables.Rd'
── Building articles ────────────────────────────────────────────────────────────────────────────────────────
Reading 'vignettes/basic-usage.Rmd'
Writing 'articles/basic-usage.html'
Reading 'vignettes/creating-datasets.Rmd'
Writing 'articles/creating-datasets.html'
══ DONE ═════════════════════════════════════════════════════════════════════════════════════════════════════
── Previewing site ──────────────────────────────────────────────────────────────────────────────────────────
> devtools::check()
Updating tradestatistics documentation
Writing NAMESPACE
Loading tradestatistics
Writing NAMESPACE
── Building ────────────────────────────────────────────────────────────────────────────── tradestatistics ──
Setting env vars:
● CFLAGS    : -Wall -pedantic -fdiagnostics-color=always
● CXXFLAGS  : -Wall -pedantic -fdiagnostics-color=always
● CXX11FLAGS: -Wall -pedantic -fdiagnostics-color=always
─────────────────────────────────────────────────────────────────────────────────────────────────────────────
✔  checking for file ‘/home/pacha/github/tradestatistics/DESCRIPTION’ ...
─  preparing ‘tradestatistics’:
✔  checking DESCRIPTION meta-information ...
─  installing the package to build vignettes
✔  creating vignettes (16.6s)
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
   Removed empty directory ‘tradestatistics/data-raw’
   Removed empty directory ‘tradestatistics/docs’
─  looking to see if a ‘data/datalist’ file should be added
─  building ‘tradestatistics_0.1.tar.gz’
   
── Checking ────────────────────────────────────────────────────────────────────────────── tradestatistics ──
Setting env vars:
● _R_CHECK_CRAN_INCOMING_USE_ASPELL_: TRUE
● _R_CHECK_CRAN_INCOMING_REMOTE_    : FALSE
● _R_CHECK_CRAN_INCOMING_           : FALSE
● _R_CHECK_FORCE_SUGGESTS_          : FALSE
── R CMD check ──────────────────────────────────────────────────────────────────────────────────────────────
─  using log directory ‘/tmp/RtmpoNTgn8/tradestatistics.Rcheck’
─  using R version 3.5.2 (2018-12-20)
─  using platform: x86_64-pc-linux-gnu (64-bit)
─  using session charset: UTF-8
─  using options ‘--no-manual --as-cran’
✔  checking for file ‘tradestatistics/DESCRIPTION’
─  checking extension type ... Package
─  this is package ‘tradestatistics’ version ‘0.1’
─  package encoding: UTF-8
✔  checking package namespace information ...
✔  checking package dependencies (1.3s)
✔  checking if this is a source package
✔  checking if there is a namespace
✔  checking for executable files ...
✔  checking for hidden files and directories
✔  checking for portable file names
✔  checking for sufficient/correct file permissions ...
✔  checking serialization versions
✔  checking whether package ‘tradestatistics’ can be installed (1.4s)
✔  checking installed package size ...
✔  checking package directory ...
✔  checking ‘build’ directory
✔  checking DESCRIPTION meta-information ...
✔  checking top-level files
✔  checking for left-over files
✔  checking index information ...
✔  checking package subdirectories ...
✔  checking R files for non-ASCII characters ...
✔  checking R files for syntax errors ...
✔  checking whether the package can be loaded (394ms)
✔  checking whether the package can be loaded with stated dependencies (358ms)
✔  checking whether the package can be unloaded cleanly (370ms)
✔  checking whether the namespace can be loaded with stated dependencies (361ms)
✔  checking whether the namespace can be unloaded cleanly (388ms)
✔  checking loading without being on the library search path (437ms)
✔  checking dependencies in R code (395ms)
✔  checking S3 generic/method consistency (870ms)
✔  checking replacement functions (385ms)
✔  checking foreign function calls (394ms)
✔  checking R code for possible problems (2.3s)
✔  checking Rd files ...
✔  checking Rd metadata ...
✔  checking Rd line widths ...
✔  checking Rd cross-references ...
✔  checking for missing documentation entries (407ms)
✔  checking for code/documentation mismatches (1.2s)
✔  checking Rd \usage sections (956ms)
✔  checking Rd contents ...
✔  checking for unstated dependencies in examples ...
✔  checking contents of ‘data’ directory
✔  checking data for non-ASCII characters ...
✔  checking data for ASCII and uncompressed saves ...
✔  checking installed files from ‘inst/doc’ ...
✔  checking files in ‘vignettes’ ...
✔  checking examples (976ms)
✔  checking for unstated dependencies in ‘tests’ ...
─  checking tests ...
✔  Running ‘testthat.R’
✔  checking for unstated dependencies in vignettes ...
✔  checking package vignettes in ‘inst/doc’ ...
✔  checking re-building of vignette outputs (30.6s)
   
   
── R CMD check results ───────────────────────────────────────────────────────────── tradestatistics 0.1 ────
Duration: 48.7s

0 errors ✔ | 0 warnings ✔ | 0 notes ✔
```

**Comment by maelle** [comment]
@pachamaltese I've just re-run R CMD Check but not via `devtools::check()` and the warning disappeared :open_mouth: so it was most probably a false positive, sorry about that.

**Comment by maelle** [comment]
@emilyriederer @mpadge thanks for accepting to review this package! :smiley_cat:  Your reviews are due on 2019-01-28. As a reminder here are our [reviewer guide](https://ropensci.github.io/dev_guide/reviewerguide.html) and [review template](https://ropensci.github.io/dev_guide/reviewtemplate.html).

**Comment by pachamaltese** [comment]
@emilyriederer Thanks a lot for this *huge* review!! I have a lot of homework now

**Comment by emilyriederer** [comment]
@pachamaltese Thanks for the package! It was very fun to play with. I hope it's helpful, and sorry for the length! All of the items are really small, and some I debated if they were worth mentioning. Personally, my struggle is always with not enough feedback, so I thought I'd share all my thoughts and you could do with them what you will 😄

**Comment by maelle** [comment]
Many thanks for your thorough review @emilyriederer! :smile_cat:

**Comment by mpadge** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- [x] As the reviewer I confirm that there are no [conflicts of interest](#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s)** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions in R help
- [x] **Examples** for all exported functions in R Help that run successfully locally
- [ ] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

- There are currently no contribution guidelines, nor a "CONTRIBUTING" file. As noted by @emilyriederer, the "BugReports" field in the "DESCRIPTION" needs updating.

##### For packages co-submitting to JOSS

- [ ] The package has an **obvious research application** according to [JOSS's definition](http://joss.theoj.org/about#submission_requirements)

The package contains a `paper.md` matching [JOSS's requirements](http://joss.theoj.org/about#paper_structure) with:

- [ ] **A short summary** describing the high-level functionality of the software
- [ ] **Authors:**  A list of authors with their affiliations
- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience.
- [ ] **References:** with DOIs for all those that have one (e.g.  papers, datasets, software).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [x] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

#### Final approval (post-review)

- [ ] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**

Estimated hours spent reviewing: 4.5

---

### Review Comments

Much of the current review does little more than reiterate the phenomenally detailed review of @emilyriederer, to whom I am also indebted for saving me much of the work, and for giving me a very helpful overview of the package!

One of my motivations in reviewing this pacakge was to gain hands-on insight into how @pachamaltese had overcome the server-side problems that plagued the [previous review process](https://github.com/ropensci/software-review/issues/217). The solution is amazingly impressive, and the documentation provided on the [`tradestatistics.io` website](https://tradestatistics.io) goes miles beyond what this package alone might suggest. That is not a problem at all, but some of my suggestions below, and many of the concerns of @emilyriederer, could be very easily allayed by providing links in the documentation to relevant pages of the fantastic [tradestatistics.github.io/documentation](https://tradestatistics.github.io/documentation).

(An immediate example is the central table of [data nomenclature and availability](https://tradestatistics.github.io/documentation/data-availability.html) which is so clear and simple that it could be helpfully reproduced in the documentation for the R package.)

One other general comment, again and as always as already indicated by @emilyriederer, concerns the names of data sets and functions, which it seems fair to suggest are too general, and would be better with, for example, package-specific prefixes (see below). In this regard, I note that the whole [`tradestatistics.io`](https://tradestatistics.io) endeavour seems already set up to be -- or at least to deserve to become -- a mini-empire and go-to place for this stuff. And yet the Observatory of Economic Complexity is not the only source of such data, with arguably an equally rich source being the [OECD](http://www.oecd.org/trade/). I for one could definitely see a case for [`tradestatistics.io`](https://tradestatistics.io) offering a unified interface for accessing such data in general. Please note that I am no way suggesting that ought to be incorporated within the package at the current point in time, and certainly not as any requirement of this review process! I merely mention it as worth thinking about in terms of current nomenclature, because it offers a number of possibilities including,

1. Prefix data and functions with, for example, `oec_get_data()`, and so allow future extension to, for example, `oecd_get_data()`.
2. Provide a general prefix and general functions which sources specified by parameters, `tradestats_get_data(..., source = "oecd")` (or whatever prefix might be desired).

## General style

To continue on the above point, the names of functions and internal data sets are arguably somewhat overly generic, as is much of the function documentation.  @pachamaltese would be advised to keep in mind that this package may be loaded in a workspace with many other packages, and `?counties` simply tells me "Provides details about countries". This should be contextualised to be helpful - "Provides details about countries included in the Observatory of Economic Complexity data retreived by tradestatistics ..." or similar. Analagous arguments apply to most documentation. These are trivial but important fixes.

Function names suffer similar lack of specificity, as already emphasised by @emilyriederer. For example, `rdocumentation.org` suggests 109 functions named `get_data()` in various CRAN packages. Examples of package specific prefixes are [`sf`](https://cran.r-project.org/package=sf) with it's uniform `st_*` functions, or [`piggyback`](https://github.com/ropensci/piggyback) with `pb_*`.  Note, however, that `ts_` is used for the [`tsbox`](https://github.com/christophsax/tsbox) package, so likely best to avoid that.


## README

While I concur with the sentiments of @emilyriederer that the diagram at the outset of the README is uniquely cool, I note that most of the text is a pasted GPL-3 license which is bundled with R and so strictly not needed there.  Github repos are a separate issue, and [Best Practice](https://github.com/coreinfrastructure/best-practices-badge) currently states,

> The project MUST post the license(s) of its results in a standard location in their source repository ... E.g., as a top-level file named LICENSE or COPYING.  License filenames MAY be followed by an extension such as ".txt" or ".md". Note that this criterion is only a requirement on the source repository. You do NOT need to include the license file when generating something from the source code (such as an executable, package, or container). For example, when generating an R package for the Comprehensive R Archive Network (CRAN), follow standard CRAN practice: if the license is a standard license, use the standard short license specification (to avoid installing yet another copy of the text) and list the LICENSE file in an exclusion file such as .Rbuildignore. 

Best Practice is nevertheless an optional system, and it would suffice simply to delete the GPL-3 text from README in order to free it up to contain the useful package-related stuff everybody will expect.

## get_countrycode()

While noting @emilyriederer's comment regarding default `NULL` values, I actually think these are generally much safer than having no default values and constantly checking `if (!missing (x))`, because they allow moving straight on to simple type checks for the parameters. I nevertheless note that the following is not very informative:
```{r}
get_countrycode()
#> Error in if (countryname == "all") { : argument is of length zero
```
In her words, the package "does not explain how NULL values are treated." Moreover, the following call actually returns an object:
```{r}
get_countrycode(1)
#> There is more than one match for your search. Please try again
#>        using one of these codes:
#> # A tibble: 0 x 6
#> # … with 6 variables: country_iso <chr>, country_name_english <chr>,
#> # country_fullname_english <chr>, continent_id <int>,
#> #   continent <chr>, eu28_member <int>
```
This is because `str_to_lower` returns `countyname = "1"` and the subsequent `str_detect` call then matches that on to all countries which have changed names because these have years of change listed like,
```{r}
tradestatistics::countries$country_fullname_english [233]
#> [1] "Belgium (Belgium-Luxembourg Customs Union until 1999)"
```
(See comment below for why the result shown on screen is then actually empty.) There are lots of possibilities for handling such cases, and I'm sure @pachamaltese is capable of implementing any sensible option, but I mention this as an incentive to implement a bit more thorough pre-checking of parameter types and classes. (At least ensure argument is stripped to alpha characters only and error on anything else such as numeric.)

Another example:
```{r}
get_countrycode("par")
#> There is more than one match for your search. Please try again
#>        using one of these codes:
#> # A tibble: 1 x 6
#>   country_iso country_name_english country_fullname_english continent_id continent eu28_member
#>     <chr>       <chr>                <chr>                           <int> <chr> <int>
#>   1 pry         Paraguay             Paraguay                            5   Americas            0
```
That message is misleading because there is actually only one match, contrary to the message. As with the above mismatches of "1" onto years of name changes, this is also a mismatch of "par" onto 
```{r}
  country_iso country_name_english country_fullname_english continent_id continent eu28_member
  <chr>       <chr>                <chr>    <int> <chr>           <int>
1 pry         Paraguay             Paraguay Americas            0
2 all         All Countries        All Countries (this is an alias to show several reporters/partner…
                                                  NA NA                 NA
```
.. and "par" matches on to the protracted and irrelevant "several reporters/partners ..." detail for "All Countries". The "All Countries" is then filtered from the screen dump because that filters `country_name_english` and not `country_fullname_english`, and that is certainly an inconsistency that ought to be rectified!

Also note that there will likely be workflows desiring multiple matches. This currently works, which is good:
```{r}
get_countrycode("par|arg|vene")
```
but still issues the message, which is undesirable if I actually want all of those. That will be important to keep in mind when converting from `tidyverse` to base R string and index matching. An important starting point may be to presume that any values entered for `countryname` are likely to be the first letters, and so you can at least `grep` on string beginnings, which would then remove both the match of "par" to "...several reporters/partners" and "1" to the guff about name changes. But you'll likely know better than me there, and maybe other alternatives such as partial or fuzzy matching might be preferable?

And because we're here, I repeat the concerns of @emilyriederer, via here example for `get_productcode()`, that the current code for this step uses the potentially unnecessary `tidyverse` functions:
```{r}
countrycode <- tradestatistics::countries %>%
    dplyr::filter(
           stringr::str_detect(
                      stringr::str_to_lower(!!rlang::sym("country_fullname_english")), countryname
           )
           ) %>%
    dplyr::select(!!rlang::sym("country_iso")) %>%
    purrr::as_vector()
```
while a base R equivalent in this case is actually much clearer and simpler:
```{r}
index <- grep (countryname,
               tolower (tradestatistics::countries$country_fullname_english))
tradestatistics::countries$country_iso [index]
```

Finally, I suspect issuing a `message` rather than an `error` on `length(countrycode) != 1` is an intentional design feature to prevent the latter from breaking long tidy-type operations, but would perhaps encourage @pachamaltese to consider having `errors` as these can easily be caught in a non-breaking way (for example with the [`collateral` package](https://github.com/rensa/collateral)), yet provide additional important information in such workflows. Happy to leave to authorial judgement call, but I would otherwise suggest that an inability to return a meaningful result ought to be an error.


## get_productcode

I note some potential confusion in the overlap between the `products` data set, the return object of `get_productcode()`, and that of `get_coutrycode()`. The latter does what it says and returns a single code, while `get_productcode()` returns a filtered version of `products` with the search string appended as an extra column. This function thus doesn't really do what it says, rather it returns the full product **table** matching the search item. I'd suggest here that `get_products()` would be a more appropriate name. Maybe @pachamaltese might then see a use case for a `get_productcode()` function that returned just the "commodity_code" column, although some clarification of nomenclature would then be in order as "products" are actually referred to as "commodities". This confusion between "products" and "commodities" pervades the entire package somewhat, and is likely something worth clarifying wherever possible.

I'd also see a use for string matching on `group_name` as well as `product_fullname_english`. Maybe `get_product(productname = NULL, productgroup = NULL)`?

I also don't see any specific need to append the column with the search term to the return result, although grant that may be just a reflection of my ignorance here? Nevertheless, the presence of "group_name" seems to provide an "official" classification of "type_product", and so render this redundant.


## get_data

My main concern here is precisely that of @emilyriederer, and worth quoting:

> What is the envisioned use case for get_productcode() since this information
> can't be passed into get_data()? From this package and the API documentation,
> I gather there is no support for filtering on commodity codes. Is that true? I
> bet a lot of users would be interested in filtering down to a specific
> industry.

My even stronger assertion would be that there would be very few users whose workflow would **not** be `get_data() %>% filter()`. It would be trivial to automate that by implementing the two extra arguments mentioned above,
```{r}
get_data (..., product_fullname_english = NULL, group_name = NULL)
```
but then if that's to be done, the logical and full extension may as well be implemented,
```{r}
get_data (..., product_fullname_english = NULL, commodity_code = NULL,
          group_code = NULL, group_name = NULL)
```
(Because you'll likely be using base R rather than `tidyverse` for the filtering, the parameter names need not necessarily be identical with those in the table, but diverging from that may cause confusion.)

To again paraphrase @emilyriederer, in the absence of such filtering ability, it's hard to argue the value of `get_productcode()` (-> `get_products()`?), because it can only be used through the user writing their own code to extract the desired entries, or likely even just as a visual lookup table, in which case it is no different from simply looking at the website.


## Miscellaneous

`DESCRIPTION` has `Suggests: covr`, but this should be removed - it is directly loaded in the `.travis.yml`, and not related to the package itself.


## Checklist

To aid @pachamaltese, this condenses (i think) most of the above points into a simpler checklist. Items with (O) are those I am happy to consider optional and leave the ultimate decision to @pachamaltese.

- [ ] Contributing guidelines; fix DESCR (including removing `covr`)
- [ ] Make function documentation more specific
- [ ] Make names of all functions and data sets more package-specific
- [ ] Remove License text from README (and possibly place elsehwhere), and provide a bit more package-related useful stuff.
- [ ] Fix fail cases and misleading errors for `get_countrycode()`
- [ ] (O) Convert dependency-heavy "tidy" code to base R
- [ ] `get_productcode()` -> `get_product(productname = NULL, productgroup = NULL)` or something along those lines.
- [ ] `get_data()` -> get_data (..., product_fullname_english = NULL, commodity_code = NULL, group_code = NULL, group_name = NULL)` or something along those lines
- [ ] Improve string matching in `get_countycode()`
- [ ] (O) Convert messages on `get_countycode()` failure to errors?

Thanks to @maelle and @pachamaltese for the opportunity to review this package, which has been a great pleasure! It is obviously the result of a phenomenal amount of work culminating in [`tradestatistics.io`](https://tradestatistics.io), of which this R package represents a small component. Congrats @pachamaltese on devising and implementing such an important and consummately well-executed project!

**Comment by maelle** [comment]
Many thanks for your review @mpadge! :rocket: 

@pachamaltese, now both reviews are in! :tada:

**Comment by pachamaltese** [comment]
Thanks a lot @maelle , @emilyriederer  and @mpadge 
Provided tradestatistics.io was basically an annoyed version of me trying to create a stable and open server to download trade data, I shall start doing some changes to the API so that data filtering by commodity is made in a more efficient way and without downloading data that then you will discard.

Does it work? What do you think about adding a commodity group filtering option to the API and then completing that nice to-do list from above?

**Comment by pachamaltese** [comment]
Hi @mpadge 

Once again, thanks a lot for the useful list.
To achieve some parts of the list without passing unwanted data I had to edit the API during lunch time.
Check this: https://api.tradestatistics.io/yrpc?y=2000&r=chl&p=arg&g=95
`g` is a new parameter that can be any HS group code (i.e. 01, 02, ..., 95, ... 99)
In the example above, the background writes a query like `... WHERE LEFT(commodity_code, 2) = 95`

Another new filtering option in the API is that any table with commodity column (the `g` example also applies to any table with commodity field), for example: https://api.tradestatistics.io/yc?y=1962&c=0101

With these changes I can start improving the package

**Comment by mpadge** [comment]
That's fantastic, @pachamaltese, and obviously a way more efficient solution than post-filtering. This review is become more and more interesting, with your ability to make server-side changes and then update R code. And all that "during lunch time" :plate_with_cutlery:

**Comment by emilyriederer** [comment]
@mpadge Really enjoyed reading your review! I really enjoyed hearing how you thought through it. Very interesting to see where we were aligned or to hear your perspective on the few places we differed. In particular, I do see your point on the `NULL` function defaults, so I retract my comment there. 😄 

@pachamaltese I think I somehow missed the full context when I wrote my review that you seem to completely own the API? That's awesome! Very psyched you were able to add filtering by commodity, as this will probably majorly simplify things for users.

**Comment by pachamaltese** [comment]
@emilyriederer Yes, he API is totally mine. I am not using the OEC (I worked with them for ~3 yrs and they never credited me...but dirty clothes are washed at home!), That's what I am using UN data + my own cleaning scripts. 

@emilyriederer @mdpage What do you think about adding a continent alias to the API (e.g. `yrp?y=2000&r=m-asia` where `m-asia` means "all ISO codes of countries in Asia)

**Comment by emilyriederer** [comment]
@pachamaltese That sounds awesome! I can envision that adding a lot of value for users to have an intermediate step between country-level of global data. The only think I think of these is potential confusion if the continent refers to the reporter, partner, or both.

**Comment by pachamaltese** [comment]
@emilyriederer what do you think about some "meta-variables" such as `c-am`, `c-af`, etc (i.e. "continent - americas"

**Comment by emilyriederer** [comment]
@pachamaltese Apologies on my last comment. I realize now I misread yours! I thought the content aliases were being filtered by a new variable and not being passed through `r` and `p`. So please disregard that I said before. Really do love the idea of that sort of flexibility, though, and the `c-{}` structure seems pretty intuitive to me!

**Comment by pachamaltese** [comment]
@emilyriederer @mpadge 
Hi. Sorry for the very long silence.
Today I worked on the changes. Are you ok about splitting the "get data" function in two functions?
I re-organized the data downloading part in:
1. `ots_read_from_api()`
2. `ots_create_tidy_data()`
Being the first more like an internal function, but still easy to use.

**Comment by mpadge** [comment]
Hi @pachamaltese no worries, and equal apologies my side for not responding to your last couple of questions! As for your current Q: It seems to me that `ots_create_tidy_data()` returns everything that `ots_read_from_api()` does and more, and therefore the latter is not needed at all, I'd suggest. A single function is always less confusing - see analogous discussions in the recent review of [`stats19`](https://github.com/ropensci/software-review/issues/266) [here](https://github.com/ropensci/software-review/issues/266#issuecomment-445504007) and [here](https://github.com/ropensci/software-review/issues/266#issuecomment-449010809) regarding "merge the formatting and reading step". That package also started off with 2 functions, one for downloading/returning the "raw" data, and a second one for processing/re-formatting that data to return a "tidy" equivalent. I agree with the suggestion of @daranzolin there that a single function is simply way less confusing. `stats19` ended up with a single `get_stats19()` function, and I'd suggest a similar thing here: simply revert to a single `ots_read_data()` function by directly renaming `ots_create_tidy_data()`, and removing `ots_read_from_api()` as an exported function.

**Comment by emilyriederer** [comment]
@pachamaltese Sorry for the slow reply!

@mpadge Are you recommending having fewer total functions or fewer *external* functions? I think the modularization of this function might be related to parts of my feedback that breaking out more internal helper functions might allow for more specific testing. The discussion you point to is very interesting, though, so I can definitely see pros/cons either way. Sorry, @pachamaltese if this feedback was somewhat erroneous!

**Comment by pachamaltese** [comment]
@mpadge @emilyriederer 
sorry for the slow response, in a move to make everyone happy I moved the "read from api part" to an internal function, so that the code is divided in smaller parts and the user is not exposed to a ton of functions

with this I can say all the indications, except for the "less tidyverse part", are complete (I love the tidyverse)

**Comment by mpadge** [comment]
Sounds great @pachamaltese. And with due apologies for delayed reponse @emilyriederer, I was definitely intending something along the lines of "fewer *external* functions", not in a way necessarily related to testing at all. So pretty much what @pachamaltese has now implemented. And finally does "all the indications ... are complete" suggest you now think the package is sufficiently stable for us to take a more thorough second look?

**Comment by pachamaltese** [comment]
@mpadge yes! I've edited a lot, you can get a 2nd look and after that I can do more changes if required

**Comment by mpadge** [comment]
It'll likely be later this week, but i'll get on to it asap. Thanks! :smiling_imp:

**Comment by emilyriederer** [comment]
Hey @pachamaltese - Just wanted to drop a note to touch base / apology. I started looking last weekend but didn't make it all the way through. I will try to reply by this weekend at the latest!

**Comment by pachamaltese** [comment]
@emilyriederer no worries!

**Comment by emilyriederer** [comment]
Hi @pachamaltese - I haven't read through the code yet, but I notice there are still a few typos in the documentation I might recommend clearing up:

- issues link still broken (change in Description and re-run `pkgdown::build_home()`)
- still missing top-level package documentation (can create with `usethis::use_package_doc()`
- "How to use this package" vignette
    + still calls package `oec`
    + contains broken image reference
- Personally, I think I would still recommend saving the "creating datasets" information just in an R script versus a vignette. It could confuse end-users since this is not something they should ever need to do. Totally fine if you disagree, of course, but in case this comment got lost in my massive review before, I thought I would reshare for consideration

**Comment by pachamaltese** [comment]
thanks @emilyriederer !! I fixed all of that and with respect to the last comment, I added "This vignette contains something end users should ever need to do"

**Comment by mpadge** [comment]
With due apologies to @pachamaltese and @maelle for this having taken a bit longer than I suggest above ...

### Second Review

The first thing to note is the now excellent message issued on package load, clearly explaining the terms and limitations of the data license - very happy with that! Neither `devtools::check()` nor `devools::spell_check()` raise any issues at all, so all clear there too. The `ots_` prefixes prepended to all function names are also a welcome improvement, Function documentation has also generally been considerably improved. 

### get_countrycode()

The following still need to be fixed:
```{r}
ots_country_code()
#> Error in ots_country_code() : is.character(countryname) is not TRUE
ots_country_code(1)
#> Error in ots_country_code() : is.character(countryname) is not TRUE
```
These errors are still not very informative, and would be better with a more human-centred message like "please provide a country name as a character string".

It might also be useful to include in function examples some examples of actual output from `?ots_country_code` which currently simply has:
```{r}
#' # Single match with no replacement
#' ots_country_code("Chile")
#' 
#' # Single match with replacement
#' ots_country_code("America")
#' 
#' # Double match with no replacement
#' ots_country_code("Germany")
```
The output of these calls is **not** rendered in the function help, so it would be better to have something more informative like:
```{r}
#' # Single match with no replacement
#' ots_country_code("Chile") # Returns "chl"
#' 
#' # Single match with replacement
#' ots_country_code("America") # Returns "usa"
#' 
#' # Double match with no replacement
#' ots_country_code("Germany") # Multiple matches; returns "ddr" and "deu"
```

I also note a non-namespaced use of `dplyr::filter` in the code:
```{r}
    f <- tradestatistics::ots_attributes_countries %>%
      filter(
        stringr::str_detect(
          stringr::str_to_lower(!!sym("country_name_english")),
          stringr::str_to_lower(countryname)
        )
      )
```
Should be `dplyr::filter`, not `filter`.


### products
The former lack of clarity between what are now named `ots_attributes_products` (a data set) and `ots_product_code` (a function) has largely been satisfactorily resolved, with the exception of the remaining confusion in nomenclature between product and commodity. Both the table and the function actually contain/return "commodity_code" values, not *product* codes - this confusion really ought to be resolved somehow, either by naming all columns in the table "product_code" and "product_fullname_english" or, alternatively, leaving the tables as they are and renaming both the data and function to "ots_attributes_commodities" and "ots_commodity_code".

My only further note is that is would be useful to have the function avoid the following undesirable behaviour:
```{r}
ots_product_code()
#> Error in ots_product_code() : object 'd' not found
```
I would suggest that it should in this case simply return the whole table. This could be done with a simple line at the outset:
```{r}
if (is.null(productname))
    return (tradestatistics::ots_attributes_products)
```

### ots_create_tidy_data

I merely reiterate my previous comment (with updated function name) that,

> there would be very few users whose workflow would **not** be `ots_create_tidy_data() %>% filter()`. It would be trivial to automate that by implementing ... extra arguments 

In your updated and improved form, this still translates to a suggestion to incorporate additional "commodity_code" and "commodity_fullname_english" (presuming renaming as suggested above), because, as already stated, it is otherwise hard to see any convincing reason to have the `ots_attributes_products` table and corresponding `ots_product_codes()` function. These are "helper" functions to focus queries on desired products/commodities, yet there is no direct way in the current implementation to subsequently use the help gained by these functions to directly extract the data.

The documentation Description for `ots_create_tidy_data` has "... and performs different API calls to return tidy data. and data transforming." - that should be fixed.

### Misc

I note that the documentation for the `ots_attributes_` data sets appears to have been manually entered (in `man/ots_attributes_...`) rather than auto-generated by `roxygen2`. This is just a suggestion, and perhaps @pachamaltese has a good reason to keep things that way, but just in case it helps: These can be auto-generated by simply adding the entries in an appropriate `./R` file (`R/tradestatistics-package.R` would seem a good place):
```{r}
#' ots_attributes_countries
#'
#' ...
#' @name ots_attributes_countries
#' @docType data
NULL
```
This does have the distinct advantage of putting the source for all documentation together in the one place - within the `./R` folder.

### checklist (#2)

Again, for your aid @pachamaltese:

- [ ] better examples in `?ots_country_code`
- [ ] More informative error messages for `ots_country_code()`
- [ ] Remove confusion between "product" and "commodity"
- [ ] Have `ots_product_code()` return the full table rather than error
- [ ] Implement additional `commodity/product_code` and `commodity/product_fullname_english` parameters in `ots_create_tidy_data` to implement filtering with single extra parameter.

Hope that helps! Thanks @pachamaltese for the great improvements thus far!

**Comment by pachamaltese** [comment]
- [ ] better examples in `?ots_country_code`
- [x] More informative error messages for `ots_country_code()`
- [x] Remove confusion between "product" and "commodity"
- [ ] Have `ots_product_code()` return the full table rather than error
- [ ] Implement additional `commodity/product_code` and `commodity/product_fullname_english` parameters in ots_create_tidy_data to implement filtering with single extra parameter.

**Comment by pachamaltese** [comment]
@mpadge Hi what do you mean with "product_fullname_english parameters in ots_create_tidy_data to implement filtering with single extra parameter"?

**Comment by mpadge** [comment]
I just mean a very slight modification to current code so that instead of users having to do this in current form:
``` r
x <- ots_create_tidy_data (years = 2015, reporter = "idn", partner = "nld") # 1,067 rows
x %>% dplyr::filter (grepl ("spice", x$group_name, ignore.case = TRUE)) # 20 rows
x %>% dplyr::filter (grepl ("ginger", x$product_fullname_english, ignore.case = TRUE)) # 1 row
```
where the `group_name` or `product_fullname_english` have to be manually matched by post-extraction filtering, this could simply be incorporated as an additional line at the end of the code:
``` r
ots_create_tidy_data <- function (..., group_name = NULL, product_fullname_english = NULL) {
...
    if (!is.null(group_name))
        data <- data %>% dplyr::filter (grepl (group_name, data$group_name, ignore.case = TRUE))
    if (!is.null (product_fullname_english))
        data <- data %>% dplyr::filter (grepl (product_fullname_english,
            data$$product_fullname_english, ignore.case = TRUE))
    if (nrow (data) == 0)
        message("No data returned, perhaps try ... [some kind of informative message here]")
    return (data)
}
```
That would then enable the much more intuitive interface
```
x <- ots_create_tidy_data (years = 2015, reporter = "idn", partner = "nld",
        product_fullname_english = "ginger")
```
It is of course up to you whether you might like to think about performing the filtering at the rest API level, because in this example case the amount of data finally required is > 1,000 times less. Either way, motivation is simply my impression that a very general use of these data will always be to implement such post-filtering anyway, so it would make the package more user friendly to implement this simple change.

**Comment by mpadge** [comment]
Another thought: This is just a suggestion and entirely a matter of personal style, but since you're importing `magrittr` anyway, you could also import `%<>%` and so replace all current lines like
``` r
data <- data %>% some_kind_of_verb(...)
```
with
``` r
data %<>% some_kind_of_verb(...)
```
There are definitely people who would consider the second version somewhat easier to read and understand - myself included. Entirely up to you of course ...

**Comment by pachamaltese** [comment]
@mpadge I opted for something a bit simpler. What do you think about this?
```r
ots_create_tidy_data(year = 1980:1981, reporter = "chl", partner = "arg", product = "0101")
ots_create_tidy_data(year = 1980:1981, reporter = "chl", partner = "arg", product = "01")
ots_create_tidy_data(year = 1980:1981, reporter = "chl", partner = "arg", product = "apple")
```

the second returns all the products in group 01 while the third returns all the rows that contain "apple" an product and/or group name.

**Comment by mpadge** [comment]
Yeah, that's a good solution too. My main concern was that without this kind of functionality, there was no direct link between the product/commodity tables and the actual trade volumes. The ability to at least put the product codes in is sufficient to make that direct link, and in my opinion seems likely to make the package much more directly usable by more people. Nice!

**Comment by pachamaltese** [comment]
@mpadge thanks !! I also renamed all the relevant fields to `product` (and therefore I had to create the compressed filed from tradestatistics.io again as well as updating the code and processing repo)

**Comment by mpadge** [comment]
Great stuff @pachamaltese - sorry if that was a hassle for you, but i really did think that the terminological confusion of asking for "products" and getting "commodities" was really worth resolving. Thanks!

**Comment by pachamaltese** [comment]
no worries @mpadge, actually UN Comtrade "commodity" nomenclature is confusing, if you look product tables the HS codes contain much more than commodities

**Comment by pachamaltese** [comment]
@mpadge ready! I implemented `%<>%`

**Comment by mpadge** [comment]
Great @pachamaltese, thanks in advance. I'll make time next week, and anticipate giving the all-clear then. Nice job!

**Comment by emilyriederer** [comment]
Apologies for my slow follow-up, and thanks to @mpadge and @pachamaltese for the great reviews / discussions which gave me a lot of great context.

# General

Great job, @pachamaltese ! A lot of the changes you’ve made I think have taken an already very good package to the next level. The level of documentation and the carefully crafted API should make this package quite accessible for new users to learn and use - that is great! 

# Package Maintenance

Thanks for adding more documentation / checks around which years of data are valid!

One question on package upkeep: do you plan to add additional years of trade data to your API going forward? If so, it may be awkward for a maintainer to manually change the `all(year %in% 1962:2016) != TRUE` condition in `ots_get_tidy_data()` and to update the documentation / website accordingly.

One potential solution would be to build in an API endpoint that reports out the maximum year of data contained and an internal function in the R package that can dynamically make that API call and use the output in `ots_get_tidy_data()`. 

Alternatively, this should at least somehow be well-documented for maintainers that API  changes should lead to multiple changes in the `ots_get_tidy_data()` file and the artifacts it generates.

# Documentation

- Nice changes to the vignette! I felt like the examples were a lot more illuminating / use-case motivated for me which makes it a fun and very useful read
- `ots_get_tidy_data()` has a `year` parameter which can conveniently be more than one year. However, the documentation reads "`Numeric value greater or equal to 1962 and lower of equal to 2016. Default set to NULL`" which I worry users will interpret to mean a single year. Perhaps "value or vector"  

# API

- I still wonder if it would make sense to have a consistent output type for `ots_product_code()` and `ots_country_code()`. The former returns a `tibble` regardless of 0, 1, or multiple matches where as the later sometimes returns a vector and other times a country code. Personally, as a user, the inconsistency *between* these two functions and *within* the `ots_country_code()` function is a bit confusing to me. I'm sure there's a good reason for it to be this way, but perhaps this can also somehow be highlighted in the documentation?

# Dependencies

Of course, you should use whichever of base of `tidyverse` that you think is best for you and your users! That said, I might still advocate for swapping out `stringr` functions with the base equivalents I mentioned above since that is a rather large package and you are using so few functions from it. Again though, I know this is totally a personal style thing so not trying to be annoying! 🙂 

# Implementation

I love the more helpful / informative error messages with the argument checks! ❤️ 

# Checklist 

- [ ]  Help maintainers keep date range code / documentation in sync with API
- [ ]  Consider rephrasing `year` parameter documentation for `ots_get_tidy_data()`
- [ ]  Consider output types of `ots_country_code()` (sometimes vector sometimes tibble) versus `ots_product_code()` and standardize or document

**Comment by pachamaltese** [comment]
Hi @emilyriederer !

Thanks a lot !! The only part I'm not sure is dropping `stringr` since I'm using `str_detect`, `str_to_lower`, `str_trim` and `str_squish`. I'm aware that `gsub(..., perl = TRUE)` can replace that but still I'm not sure about readability.

With respect to the other parts:

- [x] *Help maintainers keep date range code / documentation in sync with API.* Ready, I added https://api.tradestatistics.io/year_range and it's used in the package and documentation
- [x] *Consider rephrasing year parameter documentation for ots_get_tidy_data().* Ready, not just that, I also edited the code to allow multiple parameters like `ots_create_tidy_data(years = 1980, reporters = c("chl", "per"), partners = c("arg", "bra"), products = c("0101", "apple"))` that was not possible until you suggested to improve the docs, so I went to the very end.
- [x]  *Consider output types of ots_country_code() (sometimes vector sometimes tibble) versus ots_product_code() and standardize or document.* Ready, with all the editions needed in other parts of the package. Now both product and country searches shall return a tibble.

**Comment by emilyriederer** [comment]
Thanks @pachamaltese - makes sense! Whatever you think is best 🙂

Thanks for considering my other thoughts. I love these last changes that you've made! It's such a cool and unique thing that you're in charge of both the API and the wrapper package! It's really been interesting opening up the possibility of changing either to improve the experience.  Also *very* cool that not all the functions can take multiple options like that!

**Comment by pachamaltese** [comment]
Thanks a lot @emilyriederer , and of course also to @mpadge for helping to make this package better. Just blame @jbkunst that 3 yrs ago insisted that I should drop base+data.table and move to the tidyverse :P

**Comment by maelle** [comment]
:wave: @pachamaltese! Are you still making changes in response to the reviews or are you done?

**Comment by mpadge** [comment]
Sorry @maelle, the ball is currently pretty much in the hands of @emilyriederer and myself, who both have to go over the most recent changes one more time. I'll get on to that asap, but maybe not until early next week, and anticipate checking the accept box then.

**Comment by pachamaltese** [comment]
Hi Maelle
No, I completed all of the steps listed above.

**Comment by maelle** [comment]
Ok! @emilyriederer @mpadge, if you're happy with the changes implemented, can you check the box "The author has responded to my review and made changes to my satisfaction. I recommend approving this package." in your reviews?

Thanks to all of you for a very active and productive discussion in this thread!

**Comment by emilyriederer** [comment]
Box is checked on my end! Thanks, @pachamaltese , for such an cool API and package! You make me want a reason to need trade data so I can play with this even more 😄

**Comment by pachamaltese** [comment]
Thanks a lot @emilyriederer !! Your feedback gave me tons of ideas I'd never figured myself. Now I'm working in a "GUI" (a shiny dashboard) for the package. I shall release a beta later today.

**Comment by pachamaltese** [comment]
@maelle @emilyriederer @mpadge can you help me sharing this dashboard that is basically a GUI for the package? https://shiny.tradestatistics.io/
the main website is https://tradestatistics.io/

**Comment by mpadge** [comment]
Completed above and copied here for completeness:

#### Final approval (post-review)

- [x] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**

I've sent a very minor pull request of the [only outstanding issue](https://github.com/tradestatistics/tradestatistics/pull/1). Thanks so much @pachamaltese for the amazing work behind the scenes with your own server, and for the seamless package interface. I'll definitely be using this package!

**Comment by pachamaltese** [comment]
@mpadge thanks a lot!! I passed that point somewhere in the middle of the thread. I just rebuilt both vignettes and pkgdown site with your changes to the ots_create_tidy_data Rd section

**Comment by maelle** [comment]
Approved! Thanks @pachamaltese  for submitting and @emilyriederer @mpadge  for your reviews! :smile_cat: 

To-dos:
- [ ] Transfer the repo to rOpenSci's "ropensci" GitHub organization under "Settings" in your repo.  I have invited you to a team that should allow you to do so.  You'll be made admin again once you do.
- [ ] Add the rOpenSci footer to the bottom of your README
"```
[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)```"
- [ ] Fix any links in badges for CI and coverage to point to the ropensci URL. We no longer transfer Appveyor projects to ropensci Appveyor account so after transfer of your repo to rOpenSci's "ropensci" GitHub organization the badge should be `[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/ropensci/pkgname?branch=master&svg=true)](https://ci.appveyor.com/project/individualaccount/pkgname)`.
- [ ] We're starting to roll out software metadata files to all ropensci packages via the Codemeta initiative, see https://github.com/ropensci/codemetar/#codemetar for how to include it in your package, after installing the package - should be easy as running codemetar::write_codemeta() in the root of your package.

Should you want to awknowledge your reviewers in your package DESCRIPTION, you can do so by making them `"rev"`-type contributors in the `Authors@R` field (with their consent).  More info on this [here](https://ropensci.github.io/dev_guide/building.html#authorship).

Welcome aboard! We'd also love a blog post about your package, either a short-form intro to it (https://ropensci.org/tech-notes/) or long-form post with more narrative about its development. (https://ropensci.org/blog/). If you are interested, @stefaniebutland will be in touch about content and timing.

We've started putting together a gitbook with our best practice and tips, [this chapter](https://ropensci.github.io/dev_guide/collaboration.html) starts the 3d section that's about guidance for after onboarding. Please tell us what could be improved, the corresponding repo is [here](https://github.com/ropensci/dev_guide).

**Comment by pachamaltese** [comment]
Dear @maelle 

I've completed these steps:

- [x] Transfer the repo to rOpenSci's "ropensci" GitHub organization under "Settings" in your repo. I have invited you to a team that should allow you to do so. You'll be made admin again once you do.
- [x] Add the rOpenSci footer to the bottom of your README " [![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)"
- [x] Fix any links in badges for CI and coverage to point to the ropensci URL. We no longer transfer Appveyor projects to ropensci Appveyor account so after transfer of your repo to rOpenSci's "ropensci" - [x] GitHub organization the badge should be [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/ropensci/pkgname?branch=master&svg=true)](https://ci.appveyor.com/project/individualaccount/pkgname).
- [x] We're starting to roll out software metadata files to all ropensci packages via the Codemeta initiative, see https://github.com/ropensci/codemetar/#codemetar for how to include it in your package, after installing the package - should be easy as running codemetar::write_codemeta() in the root of your package.

Also important:
- [ ] I also updated all links in DESCRIPTION, Readme and vignettes that point to tradestatistics.github.io/* to ropensci.github.io. Is that ok?
- [ ] Besides it, I kept Joshua, Amanda and Jorge as ctb provided they gave tons of feedback on previous iterations. Shall I keep them? 
- [x] UN Comtrade is be credited as dtc.
- [ ] Shall I add Emily and Mark as ctb?

**Comment by maelle** [comment]
I don't see your repo under the ropensci org, are you sure you've transferred it?

**Comment by maelle** [comment]
Yes to keeping previous contributors over time, and regarding reviewers, you should ask them :wink:

**Comment by pachamaltese** [comment]
@emilyriederer @mpadge are you ok with being added as ctb for all the feedback?

**Comment by pachamaltese** [comment]
not yet, I am clearing ctb before pushing again before moving the repository
> I don't see your repo under the ropensci org, are you sure you've transferred it?

**Comment by pachamaltese** [comment]
@maelle ready, repository is transferred

**Comment by pachamaltese** [comment]
@maelle The only thing I cannot change is the repo description in github:
![image](https://user-images.githubusercontent.com/10091065/55490143-a4bc2980-5609-11e9-97a3-ad1737c47c25.png)
It should say https://ropensci.github.io/tradestatistics but I cannot edit that header

**Comment by pachamaltese** [comment]
I also can't change "R package to access our API" to "R package to access Open Trade Statistics API" since the repo has moved

**Comment by maelle** [comment]
@pachamaltese now you're admin again :grin:

**Comment by maelle** [comment]
Speaking of `pkgdown` feel free to use the brand-new rOpenSci style https://ropensci.github.io/rotemplate/ (feedback on the package and its guidance very welcome).

**Comment by emilyriederer** [comment]
Wow, thank you @pachamaltese ! I don't really feel that I deserve to be considered a `ctb` when all of the hard work was yours. Opinions are cheap 🙂 But very honored to be included and really appreciate it!

**Comment by maelle** [comment]
By the way @pachamaltese we recommend the role "rev" for reviewers since it's more specific than "ctb". Folks can also be both, of course.

To easily add folks to DESCRIPTION I recommend e.g. `desc::desc_add_author_gh("emilyriederer", role = "rev")`.

**Comment by maelle** [comment]
By the way @pachamaltese we recommend the role "rev" for reviewers since it's more specific than "ctb". Folks can also be both, of course.

To easily add folks to DESCRIPTION I recommend e.g. `desc::desc_add_author_gh("emilyriederer", role = "rev", comment = "reviewed the package for rOpenSci, see https://github.com/ropensci/onboarding/issues/274"))` (I'm planning on adding a better helper to `rodev` that'd add the comment automatically).

**Comment by pachamaltese** [comment]
thanks @maelle !!
I was trying to add
```

      person(given = "Mark",
             family = "Padgham",
             role = "rev",
             comment = c("reviewed the package for rOpenSci, see https://github.com/ropensci/onboarding/issues/274",
                         ORCID = "0000-0003-2172-5265")),
```
but comment is limited to a single string, probably I can create a PR to ease that

**Comment by maelle** [comment]
Something like

```r
desc::desc_add_author(given = "Mark",
                      family = "Padgham",
                      role = "rev",
                      comment = "blabla", orcid = "orcidid")
```
works although the comment section looks ugly.

**Comment by mpadge** [comment]
Thanks @pachamaltese for adding my name - much appreciated! You've done an amazing job with the package. I'm looking forward to reading about it in the rOpenSci blog.

**Comment by pachamaltese** [comment]
Hi @stefaniebutland ! Of course I'd like to write a long-form post. My email is mvargas@dcc.uchile.cl and I'm also in ROpenSci slack to be contacted.
> Welcome aboard! We'd also love a blog post about your package, either a short-form intro to it (https://ropensci.org/tech-notes/) or long-form post with more narrative about its development. (https://ropensci.org/blog/). If you are interested, @stefaniebutland will be in touch about content and timing.

**Comment by stefaniebutland** [comment]
Great to hear! I'll follow up by email
